src/v2/SoftLedgerAPI.ts

```diff
@@ -0,0 +1,477 @@
+	private instance: AxiosInstance;
+	constructor(private options: SoftledgerConnectionOptions) {
+		this.logger = options.logger || NULL_LOGGER;
+		this.cache = options.cache;
+	}
+
+	// Get a new token from the Auth service via a auth request.
+	private async getNewAuth(): Promise<SoftledgerConnectionAuthResponse> {
+		return (await axios.post(this.options.authUrl, this.options.auth0Options))?.data;
+	}
+
+	private async buildInstance(): Promise<void> {
+		this.logger.info(this.options);
+		this.instance = axios.create({ baseURL: this.options.url });
+		this.instance.defaults.headers.common['Content-Type'] = 'application/json';
+		await this.setAuth();
+		if (this.options.refreshAuth === true) {
+			createAuthRefreshInterceptor(this.instance, async () => await this.setAuth(true));
+		}
+	}
+
+	private async getToken(ignoreCache: boolean = false): Promise<string> {
+		if (!ignoreCache && !!this.cache) {
+			const token = await this.cache.get();
+			if (!!token) return token;
+		}
+
+		const authData = await this.getNewAuth();
+		if (!!this.cache) await this.cache.set(authData);
+
+		return authData.access_token;
+	}
+
+	private async setAuth(ignoreCache: boolean = false): Promise<void> {
+		const token = await this.getToken(ignoreCache);
+		this.instance.defaults.headers.common['Authorization'] = `Bearer ${token}`;
+	}
+
+	private logResponse(resp: AxiosResponse, code: number = 200, message: string = 'OK'): void {
+		const { url, method, data, params } = resp?.config || {};
+		this.logger.info(`${method} ${url} ${params} ${data}: ${code} ${message}`);
+		this.logger.verbose({ url, method, data, params, responseData: resp.data, code, message });
+	}
+
+	private logError(err: AxiosError): void {
+		const { url, method, data, params } = err.response?.config || {};
+		const { code, message } = err;
+		this.logger.info(`${method} ${url} ${params} ${data}: ${err.code} ${err.message}`);
+		this.logger.verbose({ url, method, data, params, responseData: null, code, message });
+	}
+
+	private async query<ReturnType>(cb: (ax: AxiosInstance) => Promise<AxiosResponse<ReturnType>>, options?: SoftLedgerSDKOptions): Promise<ReturnType> {
+		try {
+			const inst = await this.getInstance();
+			const resp = await cb(inst);
+			delete resp?.config?.headers?.Authorization;
+			this.logResponse(resp);
+			return resp.data;
+		} catch (e) {
+			if (options?.ifExists === true && e?.response?.status === 404) {
+				this.logResponse(e.resp, 404, 'Not Found (ifExists = TRUE)');
+				return null;
+			}
+			delete e?.response?.config?.headers?.Authorization;
+			this.logError(e);
+			throw e.toJSON();
+		}
+	}
+
+	private async getOne<T>(entity: Entity, id: NumericId, options: SoftLedgerSDKOptions): Promise<T> {
+		return this.query<T>((i) => i.get(`/${entity}/${id}`), options);
+	}
+
+	private async _getAll<T>(path: string, options: SoftledgerGetRequest = {}): Promise<Array<T>> {
+		let limit = options.limit || 1000;
+		const data: Array<T> = [];
+		while (true) {
```

Comment:
while statements make me a little nervous for infinite looping, but I expect this will work fine